lib/Authentication/AuthTokens/IAMAuthToken.php

```diff
@@ -0,0 +1,132 @@
+<?php
+namespace org\gocdb\security\authentication;
+require_once __DIR__.'/../IAuthentication.php';
+
+/*
+ * Copyright (C) 2015 STFC
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * AuthToken for use with IRIS IAM.
+ *
+ * @see IAuthentication
+ * @author Sarah Byrne
+ */
+class IAMAuthToken implements IAuthentication {
+
+    private $userDetails = null;
+    private $authorities = array();
+    private $principal;
+
+      public function __construct() {
+       $this->getAttributesInitToken();
+    }
+
+    /**
+     * {@see IAuthentication::eraseCredentials()}
+     */
+    public function eraseCredentials() {
+
+    }
+
+    /**
+     * {@see IAuthentication::getAuthorities()}
+     */
+    public function getAuthorities() {
+       return $this->authorities;
+    }
+
+    /**
+     * {@see IAuthentication::getCredentials()}
+     * @return string An empty string as passwords are not used by this token.
+     */
+    public function getCredentials() {
+        return ""; // none used in this token, handled by SSO/SAML
+    }
+
+    /**
+     * A custom object used to store additional user details.
+     * Allows non-security related user information (such as email addresses,
+     * telephone numbers etc) to be stored in a convenient location.
+     * {@see IAuthentication::getDetails()}
+     *
+     * @return Object or null if not used
+     */
+    public function getDetails() {
+        return $this->userDetails;
+    }
+
+    /**
+     * {@see IAuthentication::getPrinciple()}
+     * @return string unique principle string of user
+     */
+    public function getPrinciple() {
+       return $this->principal;
+    }
+    
+    
+    
+    private function getAttributesInitToken(){
+        if(isset($_SERVER['OIDC_access_token'])){
+            $this->principal = $_SERVER["REMOTE_USER"];
+            $this->userDetails = array('AuthenticationRealm' => array('IRIS IAM - OIDC'));
+            //Don't allow access if user only has a local account on IRIS
+            if(strpos($_SERVER['OIDC_CLAIM_groups'], "localAccounts")===false){
+            }else{
+                die('You must login via your organisation on IAM to gain access to this site.');
+            }
+            //Don't allow access unless user is a member of the IRIS gocdb group
+            if(strpos($_SERVER['OIDC_CLAIM_groups'], "gocdb")===false){
+                die('You do not belog to the correct group(s) to gain access to this site.');
```

Comment:
```suggestion
                die('You do not belong to the correct group(s) to gain access to this site.');
```
Could this be rephrased to help the user? Something: "Contact blah blah to join blah blah groups to access this site."

Reply:
It's controlled via IAM, but would probably want to be more generic and so point you to a GOCDB admin rather than an IAM/OIDC admin?

Reply:
Or maybe just point the user back to IAM to request group membership?

```suggestion
                die('You do not belong to the correct group to gain access to this site. Please visit iris-iam.stfc.ac.uk to request to join the gocdb group.');
```

Reply:
I just thought that because you have `group(s)` it may not be as simple as requesting to join the GOCDB group? Depends if it's one group for everything or you are using subgroups to protect specific access further.

If it is one group, then not sure it's fully necessary, but to me it feels good to say it will need reviewing - ie not an instant process. Also aware it's now a very long message though.

```suggestion
                die('You do not belong to the correct group(s) to gain access to this site. Please visit iris-iam.stfc.ac.uk and submit a membership request to the GOCDB group. This shall be reviewed by a GOCDB admin.');
```

Reply:
It said group(s) because the gocdb group and the localAccounts  group were originally handled together, but it is actually now just the one group called gocdb being referred to here

Reply:
I think the long message is okay, removing the plural.

```suggestion
                die('You do not belong to the correct group to gain access to this site. Please visit iris-iam.stfc.ac.uk and submit a membership request to the GOCDB group. This shall be reviewed by a GOCDB admin.');
```

Reply:
Sorry to nit-pick but, is that submit a request **to** the group or **for** (membership of) the group? Suggest maybe : 
`... submit a request for membership of the GOCDB group for review by an administrator.`  
Is that what's meant?

Reply:
It's not nit-picking, but definitely testing my grammar skills. I think this is less confusing.

```suggestion
                die('You do not belong to the correct group to gain access to this site. Please visit iris-iam.stfc.ac.uk and submit a request to join the GOCDB group. This shall be reviewed by a GOCDB admin.');
```